microclient.go

```diff
@@ -3,13 +3,69 @@ package microclient
 
 import (
 	"context"
+	"fmt"
+	"strconv"
 
 	"github.com/go-resty/resty"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/giantswarm/microerror"
+)
+
+const (
+	prometheusNamespace = "microclient"
+)
+
+var (
+	request = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: prometheusNamespace,
+			Name:      "request",
+			Help:      "Histogram for requests.",
+		},
+		[]string{"url"},
+	)
+
+	requestError = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: prometheusNamespace,
+			Name:      "request_error_count",
+			Help:      "Counters for errors during requests.",
+		},
+		[]string{"url", "error_type", "error_message"},
+	)
+
+	responseStatus = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: prometheusNamespace,
+			Name:      "response_status_count",
+			Help:      "Counters for response status codes.",
+		},
+		[]string{"url", "code"},
+	)
 )
 
+func init() {
+	prometheus.MustRegister(request)
+	prometheus.MustRegister(requestError)
+	prometheus.MustRegister(responseStatus)
+}
+
 // Do takes a resty request function and executes it, adding metrics.
-func Do(ctx context.Context, requestFunc func(string) (resty.Response, error)) (resty.Response, error) {
-	// TODO: add metrics.
-	// TODO: actually perform the request.
-	return resty.Response{}, nil
+func Do(ctx context.Context, requestFunc func(string) (*resty.Response, error), url string) (*resty.Response, error) {
```

Comment:
IMO the interface should simply be this. 

```
Do(ctx context.Context, requestFunc func() (error)) error
```

There is actually no need to deal with the URL nor response in here. If we need the response for instrumentation we should rather use the native types. Then it should be rather something like this.

```
Do(ctx context.Context, requestFunc func() (http.Response, error)) error
```

This makes the abstraction more general and open and does not require internal knowledge of resty or whatever for the wrapper. The wrapper also is not prone to abuse the response type and cannot create weird side effects.

Reply:
We talked. Joe is glad to do it more right later on in case there is a need. I cannot provide a real argument right now.